File: python-ai/services/AIWatcherService.py
"""
AI Watcher Service
Monitors file changes and triggers re-indexing
"""

import os
from typing import Dict
from watchdog.observers import Observer
from watchdog.events import FileSystemEventHandler
from config.aiConfig import WATCHER_CONFIG
from utils.fileUtils import should_process_file, calculate_file_hash
from utils.logger import get_logger

logger = get_logger(__name__)


class FileChangeHandler(FileSystemEventHandler):
    """Handler for file system events"""

    def __init__(self, on_file_change):
        self.on_file_change = on_file_change
        self.debounce_timers = {}

    def on_modified(self, event):
        """Handle file modification"""
        if not event.is_directory and should_process_file(event.src_path):
            logger.debug(f"File modified: {event.src_path}")
            self.on_file_change(event.src_path, 'modified')

    def on_created(self, event):
        """Handle file creation"""
        if not event.is_directory and should_process_file(event.src_path):
            logger.debug(f"File created: {event.src_path}")
            self.on_file_change(event.src_path, 'created')

    def on_deleted(self, event):
        """Handle file deletion"""
        if not event.is_directory and should_process_file(event.src_path):
            logger.debug(f"File deleted: {event.src_path}")
            self.on_file_change(event.src_path, 'deleted')


class AIWatcherService:
    """Service for monitoring file changes"""

    def __init__(self):
        self.watchers = {}
        self.observer = None
        self.debounce_timers = {}

    def start_watching(self, project_path: str):
        """Start watching project directory"""
        if not WATCHER_CONFIG['enabled']:
            logger.debug("Watcher is disabled")
            return

        if project_path in self.watchers:
            logger.debug(f"Already watching project: {project_path}")
            return

        logger.info(f"Starting file watcher: {project_path}")

        try:
            if not self.observer:
                self.observer = Observer()
                self.observer.start()

            handler = FileChangeHandler(self.handle_file_change)
            watch = self.observer.schedule(handler, project_path, recursive=True)
            self.watchers[project_path] = watch

            logger.info(f"File watcher started: {project_path}")
        except Exception as e:
            logger.error(f"Error starting watcher: {project_path}, {str(e)}")

    def stop_watching(self, project_path: str):
        """Stop watching project"""
        watch = self.watchers.get(project_path)
        if watch:
            self.observer.unschedule(watch)
            del self.watchers[project_path]
            logger.info(f"Stopped watching project: {project_path}")

    def stop_all_watchers(self):
        """Stop all watchers"""
        if self.observer:
            self.observer.stop()
            self.observer.join()
            self.watchers.clear()
            logger.info("Stopped all file watchers")

    def handle_file_change(self, file_path: str, event_type: str):
        """Handle file change event"""
        logger.debug(f"File change detected: {file_path}, type: {event_type}")

        # Debounce the processing
        debounce_key = file_path
        if debounce_key in self.debounce_timers:
            import threading
            self.debounce_timers[debounce_key].cancel()

        def process_change():
            self.process_file_change(file_path, event_type)
            if debounce_key in self.debounce_timers:
                del self.debounce_timers[debounce_key]

        import threading
        timer = threading.Timer(WATCHER_CONFIG['debounce_ms'] / 1000, process_change)
        self.debounce_timers[debounce_key] = timer
        timer.start()

    def process_file_change(self, file_path: str, event_type: str):
        """Process file change"""
        try:
            logger.info(f"Processing file change: {file_path}, type: {event_type}")

            if event_type == 'deleted':
                logger.info(f"File deleted: {file_path}")
            else:
                # Re-ingest the file
                file_hash = calculate_file_hash(file_path)
                logger.info(f"File hash: {file_path}, hash: {file_hash}")

        except Exception as e:
            logger.error(f"Error processing file change: {file_path}, {str(e)}")

    def get_status(self) -> Dict:
        """Get watcher status"""
        return {
            'enabled': WATCHER_CONFIG['enabled'],
            'watching_projects': len(self.watchers),
            'projects': list(self.watchers.keys()),
        }

File: backend/ai/services/AIWatcherService.js
/**
 * AI Watcher Service
 * Monitors file changes and triggers re-indexing
 */

const fs = require('fs');
const path = require('path');
const AIIngestService = require('./AIIngestService');
const AIEmbeddingService = require('./AIEmbeddingService');
const AILogger = require('../utils/logger');
const AI_CONFIG = require('../config/aiConfig');
const { calculateFileHash, detectProjectType } = require('../utils/fileUtils');
const { getExistingTrainingModules, resolveModuleLabel, isAllowedTrainingPath } = require('../config/proposalForgeModules');

const logger = new AILogger('WatcherService');

class AIWatcherService {
  constructor() {
    this.watchers = new Map();
    this.ingestService = new AIIngestService();
    this.embeddingService = new AIEmbeddingService();
    this.debounceTimers = new Map();
    this.debounceMs = AI_CONFIG.watcher.debounceMs;
  }

  /**
   * Start watching project directories
   */
  startWatching(projectPath) {
    if (!AI_CONFIG.watcher.enabled) {
      logger.debug('Watcher is disabled');
      return;
    }

    if (this.watchers.has(projectPath)) {
      logger.debug('Already watching project', { projectPath });
      return;
    }

    logger.info('Starting file watcher', { projectPath });

    try {
      const watcher = fs.watch(projectPath, { recursive: true }, (eventType, filename) => {
        this.handleFileChange(projectPath, filename, eventType);
      });

      this.watchers.set(projectPath, watcher);
    } catch (err) {
      logger.error('Error starting watcher', { projectPath, error: err.message });
    }
  }

  /**
   * Stop watching project
   */
  stopWatching(projectPath) {
    const watcher = this.watchers.get(projectPath);
    if (watcher) {
      watcher.close();
      this.watchers.delete(projectPath);
      logger.info('Stopped watching project', { projectPath });
    }
  }

  /**
   * Stop all watchers
   */
  stopAllWatchers() {
    for (const [projectPath, watcher] of this.watchers.entries()) {
      watcher.close();
      logger.info('Stopped watching project', { projectPath });
    }
    this.watchers.clear();
  }

  /**
   * Handle file change event
   */
  handleFileChange(projectPath, filename, eventType) {
    const filePath = path.join(projectPath, filename);

    // Skip if file doesn't exist or is in excluded directories
    if (!fs.existsSync(filePath)) {
      return;
    }

    // Check if file should be watched
    if (!this.shouldWatchFile(filePath) || !isAllowedTrainingPath(filePath)) {
      return;
    }

    logger.debug('File change detected', { filePath, eventType });

    // Debounce the processing
    const debounceKey = filePath;
    if (this.debounceTimers.has(debounceKey)) {
      clearTimeout(this.debounceTimers.get(debounceKey));
    }

    const timer = setTimeout(() => {
      this.processFileChange(projectPath, filePath);
      this.debounceTimers.delete(debounceKey);
    }, this.debounceMs);

    this.debounceTimers.set(debounceKey, timer);
  }

  /**
   * Check if file should be watched
   */
  shouldWatchFile(filePath) {
    const filename = path.basename(filePath);
    const dirname = path.dirname(filePath);

    // Check if in watched paths
    const isInWatchPath = AI_CONFIG.watcher.watchPaths.some(watchPath => {
      return filePath.includes(watchPath) || dirname.includes(watchPath);
    });

    if (!isInWatchPath) {
      return false;
    }

    // Exclude certain files
    const excludePatterns = [
      /\.log$/,
      /\.tmp$/,
      /\.swp$/,
      /\.swo$/,
      /~$/,
      /node_modules/,
      /\.git/,
      /build/,
      /dist/,
    ];

    for (const pattern of excludePatterns) {
      if (pattern.test(filePath)) {
        return false;
      }
    }

    return true;
  }

  /**
   * Process file change
   */
  async processFileChange(projectPath, filePath) {
    try {
      logger.info('Processing source file change', { filePath });

      const projectName = resolveModuleLabel(filePath);
      const projectType = detectProjectType(projectPath);

      // Re-ingest the file
      await this.ingestService.ingestFile(filePath, projectName, projectPath, projectType);

      // Generate new embeddings
      const AIDocument = require('../models/AIDocument');
      const document = await AIDocument.findOne({ filepath: filePath });

      if (document) {
        const embeddings = [];

        for (const chunk of document.chunks) {
          const embedding = await this.embeddingService.generateEmbedding(chunk.content);
          embeddings.push(embedding);
        }

        // Store updated embeddings
        await this.embeddingService.storeEmbeddings(document._id.toString(), document.chunks, embeddings);

        // Mark as processed
        await this.ingestService.markEmbeddingsGenerated(document._id);
        await this.ingestService.markAsProcessed(document._id);

        logger.info('File change processed successfully', { filePath });
      }
    } catch (err) {
      logger.error('Error processing file change', { filePath, error: err.message });
    }
  }

  /**
   * Get watcher status
   */
  getStatus() {
    return {
      enabled: AI_CONFIG.watcher.enabled,
      watchingProjects: this.watchers.size,
      projects: Array.from(this.watchers.keys()),
    };
  }
}

module.exports = AIWatcherService;
